### PR TITLE
feat: pre-check actual result's data type in x:expect

### LINKED ISSUE #2089: Test on error code/messages always produce an execution error if there was no error
Hi,

I would like to point to a common problem I have many times with the `catch="yes"` implementation (#500).

The basic example of #500  was:

```xml
<x:scenario... catch="yes">
  <x:call function="conf:getConfVar">
    <x:param name="conf" xmlns="foo:bar">
      <conf>
        <variable name="foo" value="nothing"/>
      </conf>
    </x:param>
    <x:param name="varName" select="'missing'"/>
  </x:call>
  <x:expect test="$x:result?err?code='ERROR_MISSING_ENTRY'"/>
  <x:expect test="$x:result?err?description='nothing is missing in environment configuration'"/>
</x:scenario>
```

The problem I always have with this is that if the function `conf:getConfVar` does not throw an error but returns something which is not a map, I will get an execution error of the XSpec framework:

```
XPTY0004  The items on the LHS of the '?' operator must be maps or arrays; but value (document-node()) was supplied
```

What I would expect is an test failure not the termination of my XSpec process.

To achive this I always have to write something like this:

```xml
<x:expect test="
if($x:result instance of map(*))
then
$x:result?err?code='ERROR_MISSING_ENTRY'
else false()
"/>
```

This is really anoying. How are you handling this? Shouldn't be there a standard way to catch such cases?

While I'm typing, I came to a more generic idea - which is maybe also helpfull for other type-conflict cases:

```xml
<x:expect test="$x:result?err?code='ERROR_MISSING_ENTRY'" test-type="map(*)"/>
```

The rule for `<x:expect>` with `@test-type` could be: 

If `$x:result instance of {@test-type}` is **not satisfied** a test failure is reported - without evaluating `@test` (or anything else). Otherwise handle it as a regular `<x:expect>`.

An alternative would be just to change the implementation: Catch any error thrown by the evaluation of `@test` and report it as a test failure.

What do you think?

BR, Nico

### PR DESCRIPTION
**Pull request by galtm**
This pull request implements the idea suggested by @nkutsche (thanks!) in #2089 to make `x:expect` support an optional attribute that declares the data type of the actual result. This PR calls the attribute `result-type`. For a given `x:expect` element having this attribute, if `$x:result instance of {@result-type}` is false, then `x:expect` fails and `@test` is not evaluated -- which is deliberate, because an unexpected data type could lead to a fatal error if `@test` were evaluated.

### Example
```
<x:expect label="integer greater than 1" result-type="xs:integer" test="$x:result gt 1"/>
```

If the actual result is an instance of `xs:integer`, then `x:expect` works just like before: it passes if the boolean `@test` evaluates to true.

If the actual result happens to be a string for some reason, you get a failing `x:expect` instead of a fatal error.

### Usage Notes

- The `result-type` attribute is like a pre-check, not a new verification syntax in itself.
- Test execution could still halt for reasons unrelated to the result data type. Examples: `test="q:element-name"` if the prefix `q` is not declared, or `select="./child"` with no child nodes in `x:expect` to provide a context.
- If `x:expect` has no attributes or child nodes that define an expected result and also has no `@test` attribute, then the expected result is implicitly an empty sequence. To be consistent with precedent, this PR continues that behavior. If you want to verify _only_ the data type and not the value of the actual result, use something like `<x:expect test="$x:result instance of xs:integer" label="..."/>`, _not_ `<x:expect result-type="xs:integer" label="..."/>`.

### Code Review Notes

- For some description of the compiler changes, see 2cc7381106f18cc7a2b120c34a3fa08a2dc70d14
- `test/result-type-matches.xspec` has some examples of matching data types
- `test/end-to-end/cases/result-type-non-matches.xspec` has some examples of non-matching data types, where you get verification failures instead of fatal errors
- Earlier history of error handling in XSpec is in #599 and #500

Fixes #2089.

### CODE REVIEW

File: src/compiler/base/declare-variable/sequencetype-with-uqnames.xsl [hunk 1]
```diff
@@ -15,7 +15,8 @@
    -->
    <xsl:function name="x:lexical-to-UQName-in-sequence-type" as="xs:string?">
       <xsl:param name="context" as="element()"/>
-      <xsl:variable name="as-attribute-value" as="xs:string?" select="$context/@as"/>
+      <xsl:param name="attribute-local-name" as="xs:string"/>
```
Comment (galtm) [inline src/compiler/base/declare-variable/sequencetype-with-uqnames.xsl:18]: I generalized the function from #2103 so it can apply to either `@as` (relevant for #2103) or `@result-type` (relevant here).

File: src/schemas/xspec.rnc [hunk 2]
```diff
@@ -310,10 +310,14 @@ assertion =
...
 	## href and select attributes or content of the <expect> element.
 	element expect { common-attributes,
-		label, pending-attribute?, test?, selection }
+		label, pending-attribute?, test?, result-type?, selection }
 
 test = attribute test { xpath }
 
+result-type =
+	## Data type of actual result.
+	attribute result-type { xsd:string { minLength = "1" } }
```
Comment (galtm) [inline src/schemas/xspec.rnc:319]: The value is supposed to be a SequenceType (as in [instance of](https://www.w3.org/TR/xpath-31/#id-instance-of)) but I didn't express the SequenceType pattern more specifically because the `as` attribute doesn't, either.

File: test/compiler-sequencetype-with-uqnames.xspec [hunk 3]
```diff
@@ -1,124 +1,134 @@
...
-			<x:expect label="enables correct conversion to UQName"
-				select="'Q{urn:x-xspec:example}b'"/>
-		</x:scenario>
-		<x:scenario label="When prefix is declared differently on ancestor of context element">
-			<x:call function="x:lexical-to-UQName-in-sequence-type">
-				<x:param select="//element">
-					<parent xmlns:a="urn:x-xspec:example:parent">
+	<x:scenario label="Lexical QName in @as attribute">
+		<x:call function="x:lexical-to-UQName-in-sequence-type">
+			<x:param position="2" select="'as'"/>
```
Comment (galtm) [inline test/compiler-sequencetype-with-uqnames.xspec:9]: From the diff, it looks like this file had major changes, but really I was adapting the test to the function's 2nd input argument and the file's change in directory location.

File: test/end-to-end/cases/expect-empty.xspec [hunk 4]
```diff
@@ -11,6 +11,7 @@
 		<x:expect>
 			<x:label>Ditto using x:label</x:label>
 		</x:expect>
+		<x:expect label="Ditto with @result-type that permits an empty sequence" result-type="item()?" />
```
Comment (galtm) [inline test/end-to-end/cases/expect-empty.xspec:14]: Related to 3rd bullet under Usage Notes, in PR description

**Comment by cmarchand** [comment]
I'm not sure I'm able to validate this one ! This is too close to compiler for me...

**Comment by galtm** [comment]
> I'm not sure I'm able to validate this one ! This is too close to compiler for me...

No worries. I wasn't sure how close you were to the code that I changed. If you have any feedback on the design as XSpec users would see it (such as the part of the PR description _above_ "Code Review Notes"), please let me know.

**Comment by galtm** [comment]
Schematron addition:

35e16a54ce2a8bfb48984123e42baef38544fc4a adds `result-type="node()"` to the `x:expect` elements that are generated from Schematron-specific verification elements like `x:expect-assert`. Without this change, if validation causes a dynamic error, the testing process has a fatal error even if the scenario has `catch="yes"`. The error is not user-friendly, and it looks like this:
```
Type error evaluating (.) at char 157 in expression in xsl:sequence/@select on line 250 column 46 of my-test-file-compiled.xsl:
  XPTY0020  The required item type of the context item for the child axis is node(); the
  supplied value map{"err":map{(:size 6:)}} is a map
```
With this change, you get a verification failure instead of a fatal error.
